R/dalys.R

```diff
@@ -0,0 +1,180 @@
+##' Calculate dalys
+##'
+##' @title DALYs calculation
+##'
+##' @param con You can be \code{readonly} user to run this function.
+##' But if you need to import dalys for Ferrari, Li and LiST (201710gavi), you can use \code{import} user.
+##' @param dalys_src Dalys calcualtion parameters. This is currently a csv file in orderly_report i1400.
+##' We will surely move that file to model_documentation once it is complete.
+##' @param touchstone_name touchstone for which dalys are calcualted
+##' @param year_min minimal year
+##' @param year_max maximum year
+##' @param vimc_dalys_only set to be TRUE is we are only interested in Ferrari, Li and LiST (201710gavi)
+##'
+##' @export
+calculate_dalys <- function(con, dalys_src, touchstone_name, year_min, year_max, vimc_dalys_only = TRUE){
+  ## [make temporary dalys parameter table]
+  dalys_parameters <- create_dalys_parameters(con, dalys_src=dalys_src, touchstone_name, vimc_dalys_only)
+  # To validate our calculation against modeller's upload, we can restrict to specific disease
+  #dalys_parameters <- dalys_parameters[dalys_parameters$disease == 'Measles', ]
+  ## [make remaining life expectancy table] - this will take some time
+  life_table <- create_dalys_life_table(con, touchstone_name, year_min, year_max)
+
+  sets <- unique(dalys_parameters$burden_estimate_set_id)
+
+  dalys_out <- rep(list(NULL),length(sets))
+  for(i in seq_along(sets)){
+    message(sprintf("calculating dalys for burden_estimate_set %s", sets[i]))
+
+    year <- c("(SELECT burden_estimate_set, burden_estimate.country, year, age, burden_outcome, value")
+    year_range <- sprintf("AND year BETWEEN %1s AND %2s", year_min, year_max)
+    sql <- paste("SELECT burden_estimate_set, country, year, age, burden_outcome, code, value as burden, sub_condition, adjusted_weight, duration",
+                 "FROM (SELECT *",
+                 "FROM dalys_parameters",
+                 sprintf("WHERE burden_estimate_set_id = %s", sets[i]),
+                 ") AS parameters",
+                 "LEFT JOIN",
+                 year,
+                 "FROM burden_estimate",
+                 sprintf("WHERE burden_estimate_set = %s", sets[i]),
+                 year_range,
+                 ") AS burdens",
+                 "ON burdens.burden_estimate_set = parameters.burden_estimate_set_id",
+                 "AND burdens.burden_outcome = parameters.burden_outcome_id"
+    )
+    # burden_estimates
+    v <- DBI::dbGetQuery(con, sql)
+    # match dalys parameters
+    v$.code <- paste(v$country, v$year, v$age, sep="-")
+    v <- merge(v, life_table, by= ".code", all.x=TRUE)
```

Comment:
It would be good, as well as lifting this out into its own function, to see the actual core of the DALY calculation as something that does not depend on the database at all - it looks like at the core it brings a life table and some impact together - this seems very testable with a small dataset.

Reply:
Yes. Dalys calculation relates to extra information (model-specific dalys parameters from DOVE). I will make tests.